internal/shim-sgx/src/heap.rs

```diff
@@ -2,218 +2,105 @@
+        let region = Region::new(start, end);
         Self {
-            memory,
-            brk,
-            ledger: Ledger::new(Region::new(Address::new(0), memory_len)),
+            start,
+            end,
+            ledger: Ledger::new(region),
+            brk: start,
+            brk_max: start,
         }
     }
 
-    /// Get reference to the heap's memory.
-    pub fn memory(&self) -> &[Page] {
-        self.memory
+    /// Return the maximum `brk` address reached.
+    pub fn brk_max(&self) -> Address<usize, Page> {
+        self.brk_max
     }
 
     /// Returns the range of the heap area
     pub fn range(&self) -> Range<*const Page> {
-        self.memory().as_ptr_range()
-    }
-
-    fn offset(&self, addr: usize) -> Option<usize> {
-        let base = self.memory().as_ptr() as usize;
-        let ceil = base + self.memory().len() * Page::SIZE;
-
-        if base <= addr {
-            if addr < ceil {
-                return Some(addr - base);
-            } else {
-                return Some(ceil - base);
-            }
+        unsafe {
+            from_raw_parts_mut(
+                self.start.raw() as *mut Page,
+                (self.end.raw() - self.start.raw()) / Page::SIZE,
+            )
         }
-
-        None
-    }
-
-    fn offset_page_down(&self, addr: usize) -> Option<usize> {
-        let off = self.offset(addr)?;
-        Some(off / Page::SIZE)
-    }
-
-    fn offset_page_up(&self, addr: usize) -> Option<usize> {
-        self.offset_page_down(addr + Page::SIZE - 1)
+        .as_ptr_range()
     }
 
-    fn pos(&self) -> usize {
-        self.brk
-            .map(|x| x.into())
-            .unwrap_or(self.memory().as_ptr() as _)
-    }
-
-    /// Allocate heap memory to address `brk`
-    pub fn brk(&mut self, brk: usize) -> usize {
-        let old = self.offset_page_up(self.pos()).unwrap();
-        let new = match self.offset_page_up(brk) {
-            Some(page) => page,
-            None => return self.pos(),
-        };
-        match old.cmp(&new) {
-            Ordering::Less => {
-                let region = Region::new(
-                    Address::new(old * Page::SIZE),
-                    Address::new(new * Page::SIZE),
-                );
-                if self
-                    .ledger
-                    .map(region, Access::READ | Access::WRITE)
-                    .is_err()
-                {
-                    return self.pos();
-                }
-            }
-            Ordering::Greater => {
-                let region = Region::new(
-                    Address::new(new * Page::SIZE),
-                    Address::new(old * Page::SIZE),
-                );
-                self.ledger.unmap(region).unwrap();
+    /// Increase or decrease `brk` address.
+    pub fn brk(&mut self, brk: Address<usize, Page>) -> Address<usize, Page> {
+        if brk < self.start || brk >= self.end {
+            return self.brk;
```

Comment:
Shouldn't it `unmap` like before?

Reply:
No because before trimming is implemented we cannot unmap.

Reply:
Ok, then maybe add a `FIXME: ` comment

Reply:
I don't see any comment here

Reply:
added